### PR TITLE
Fix double-free in HashTable::clearAndShrink() with zero elements in it

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix double-free in HashTable::clearAndShrink() with zero elements in it

Before this patch HashTable::clearAndShrink() did this:

    destroyElements();
    clearHasZero();

While destroyElements() already destroys the zero element, so it leads to double free, like here [1].

  [1]: https://s3.amazonaws.com/clickhouse-test-reports/40003/f09e956ddf150c1aceb210348544bc28876a2d11/stateless_tests__asan__[2/4].html

The issue is only for complex types.

### CODE REVIEW

**Comment by davenger** [comment]
With this fix "has_zero" flag will not be reset to false but the zeroValue() will be destroyed.
Could this be a problem?

**Comment by azat** [comment]
>With this fix "has_zero" flag will not be reset to false but the zeroValue() will be destroyed.
>Could this be a problem?

@davenger likely. Nice catch. Fixed.

File: src/Common/HashTable/HashTable.h [hunk 1]
```diff
@@ -652,6 +658,15 @@ class HashTable : private boost::noncopyable,
                 ///   [1]: https://github.com/google/sanitizers/issues/854#issuecomment-329661378
                 __msan_unpoison(it.ptr, sizeof(*it.ptr));
             }
+
```
Comment (kitaisreal) [inline src/Common/HashTable/HashTable.h:661]: Where do we have non trivial destructible Cell in our hash tables ?

**Review by kitaisreal — CHANGES_REQUESTED** [review]

File: src/Common/HashTable/HashTable.h [hunk 2]
```diff
@@ -652,6 +658,15 @@ class HashTable : private boost::noncopyable,
...
+            /// Everything had been destroyed in the loop above, reset the flag
+            /// only, without calling destructor.
+            this->clearHasZeroFlag();
+        }
+        else
+        {
+            if (this->hasZero())
+                this->clearHasZero();
```
Comment (kitaisreal) [inline src/Common/HashTable/HashTable.h:669]: If this is branch where Cell is not trivially destructible, can't we call `clearHasZero` unconditionally ?

File: src/Common/HashTable/HashTable.h (hunk 1 above)
Comment (davenger) [reply src/Common/HashTable/HashTable.h:661]: As far as I see in this double-free report https://s3.amazonaws.com/clickhouse-test-reports/40003/f09e956ddf150c1aceb210348544bc28876a2d11/stateless_tests__asan__[2/4]/stderr.log
there is ```HashMapCell<unsigned long, DB::Array, ...``` and DB::Array has a std::vector inside

**Comment by azat** [comment]
>Performance Comparison [1/4] — 2 errors, 3 too long, 1 faster
>Performance Comparison [2/4] — 2 errors, 3 too long, 1 slower, 4 unstable
>Performance Comparison [3/4] — 6 errors, 1 too long, 7 slower
>Performance Comparison [4/4] — 2 errors, 2 too long, 1 slower, 3 unstable

- #44312

File: src/Common/HashTable/HashTable.h (hunk 1 above)
Comment (azat) [reply src/Common/HashTable/HashTable.h:661]: Yes, like here - https://github.com/ClickHouse/ClickHouse/blob/647077a4b018161413bffaf214fa000d6ee8f0b2/src/Dictionaries/HashedDictionary.h#L190

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**